crates/runtime/src/vmcontext.rs

```diff
@@ -11,24 +11,57 @@ use std::marker;
 use std::ptr::NonNull;
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::u32;
-pub use vm_host_func_context::VMHostFuncContext;
-use wasmtime_environ::DefinedMemoryIndex;
+pub use vm_host_func_context::{VMArrayCallHostFuncContext, VMNativeCallHostFuncContext};
+use wasmtime_environ::{DefinedMemoryIndex, VMCONTEXT_MAGIC};
 
-pub const VMCONTEXT_MAGIC: u32 = u32::from_le_bytes(*b"core");
+/// A function pointer that exposes the calling convention.
+pub type VMArrayCallFunction =
+    unsafe extern "C" fn(*mut VMOpaqueContext, *mut VMOpaqueContext, *mut ValRaw, usize);
+
+/// A function pointer that exposes the native calling convention.
+#[repr(C)]
+pub struct VMNativeCallFunction(pub VMFunctionBody);
+
+/// A function pointer that exposes the Wasm calling convention.
+#[repr(C)]
+pub struct VMWasmCallFunction(pub VMFunctionBody);
 
 /// An imported function.
 #[derive(Debug, Copy, Clone)]
 #[repr(C)]
 pub struct VMFunctionImport {
-    /// A pointer to the imported function body.
-    pub body: NonNull<VMFunctionBody>,
+    /// Function pointer to use when calling this imported function from Wasm.
+    ///
+    /// Most functions come with a function pointer that we can use when they
+    /// are called from Wasm. The notable exception is when we `Func::wrap` a
+    /// host function, and we don't have a Wasm compiler on hand to compile a
+    /// Wasm-to-native trampoline for the function. In this case, we leave
+    /// `wasm_call` empty until the function is passed as an import to Wasm (or
+    /// otherwise exposed to Wasm via tables/globals). At this point, we look up
+    /// a Wasm-to-native trampoline for the function in the the Wasm's compiled
+    /// module and use that fill in `VMFunctionImport::wasm_call`. **However**
+    /// there is no guarantee that the Wasm module has a trampoline for this
+    /// function's signature. The Wasm module only has trampolines for its
+    /// types, and if this function isn't of one of those types, then the Wasm
+    /// module will not have a trampoline for it. This is actually okay, because
+    /// it means that the Wasm cannot actually call this function. But it does
+    /// mean that this field needs to be an `Option` even though it is non-null
+    /// the vast vast vast majority of the time.
+    pub wasm_call: Option<NonNull<VMWasmCallFunction>>,
```

Comment:
Shouldn't this always be `Some(...)` by the time it reaches the `VMFunctionImport` phase?

Reply:
Oh good call, because the case that the above comment is talking about can't happen with function imports, only `funcref`s.